File: app/src/main/java/com/mmxb/mhelper/MainActivity.java
package com.mmxb.mhelper;

import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.Toast;

import com.mmxb.mhelper.floatwindow.FloatWindow;
import com.mmxb.mhelper.floatwindow.FloatWindowService;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }

    public void test(View view) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (Settings.canDrawOverlays(this)) {
                startFloatWindowService();
            } else {
                startActivityForResult(new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION, Uri.parse("package:" + getPackageName())), 0);
            }
        } else {
            startFloatWindowService();
        }

    }

    private void startFloatWindowService() {
        startService(new Intent(MainActivity.this, FloatWindowService.class));
    }


    @RequiresApi(api = Build.VERSION_CODES.M)
    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 0) {
            if (Settings.canDrawOverlays(this)) {
                startFloatWindowService();
            } else {
                Toast.makeText(this, "请打开悬浮窗权限", Toast.LENGTH_LONG).show();
            }
        }
    }


}

File: app/src/main/java/com/mmxb/helper/MainActivity.kt
package com.mmxb.helper

import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.support.v7.app.AppCompatActivity
import android.view.View
import com.mmxb.helper.floatwindow.FloatWindowService
import com.mmxb.helper.util.RootUtil
import com.mmxb.helper.util.ServiceUtil
import kotlinx.android.synthetic.main.activity_main.*


class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        initView()
    }

    private fun initView() {
        val isShowingFloatWindow = ServiceUtil.isServiceRunning(this, FloatWindowService::class.java.canonicalName)
        setOpenText(!isShowingFloatWindow)
    }

    fun openFloatWindow(view: View) {
        val isShowingFloatWindow = ServiceUtil.isServiceRunning(this, FloatWindowService::class.java.canonicalName)
        if (isShowingFloatWindow) {  // 关闭
            // todo
            stopService(Intent(this@MainActivity, FloatWindowService::class.java))
            setOpenText(true)
        } else {  // 打开
            // 悬浮窗权限
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.canDrawOverlays(this)) {
                switchFloatWindow.setViewChecked(false)
                val intent = Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION, Uri.parse("package:$packageName"))
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                startActivity(intent)
                return
            }
            switchFloatWindow.setViewChecked(true)


            // 无障碍服务
            if (!ServiceUtil.isAccessibilitySettingsOn()) {
                switchAccessibilityService.setViewChecked(false)
                val intent = Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS)
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                startActivity(intent)
                return
            }
            switchAccessibilityService.setViewChecked(true)

            // root权限
            val isRoot = RootUtil.isRoot(this)
            switchRoot.setViewChecked(isRoot)

            startFloatWindowService()
            setOpenText(false)
        }
    }

    private fun startFloatWindowService() {
        startService(Intent(this@MainActivity, FloatWindowService::class.java))
    }

    private fun setOpenText(open: Boolean) {
        openBtn.text = if (open) getString(R.string.open_float_window) else getString(R.string.close_float_window)
    }
}
